src/oatpp/core/data/stream/BufferStream.cpp

```diff
@@ -128,39 +128,87 @@ oatpp::String BufferOutputStream::getSubstring(v_buff_size pos, v_buff_size coun
   }
 }
 
-oatpp::v_io_size BufferOutputStream::flushToStream(OutputStream* stream) {
+v_io_size BufferOutputStream::flushBufferToStream(stream::WriteCallback* stream) {
   return stream->writeExactSizeDataSimple(m_data, m_position);
 }
 
-oatpp::async::CoroutineStarter BufferOutputStream::flushToStreamAsync(const std::shared_ptr<BufferOutputStream>& _this, const std::shared_ptr<OutputStream>& stream) {
+oatpp::async::CoroutineStarter BufferOutputStream::flushBufferToStreamAsync(const std::shared_ptr<BufferOutputStream>& _this, const std::shared_ptr<WriteCallback>& stream) {
+  return writeBufferToStreamAsync(_this, stream, _this->m_position);
+}
+
+oatpp::async::CoroutineStarter BufferOutputStream::writeBufferToStreamAsync(const std::shared_ptr<BufferOutputStream> &_this,
+                                                                            const std::shared_ptr<WriteCallback> &stream,
+                                                                            v_buff_size count) {
+  return writeBufferToStreamAsync(_this, stream, count, 0);
+}
 
+oatpp::async::CoroutineStarter BufferOutputStream::writeBufferToStreamAsync(const std::shared_ptr<BufferOutputStream> &_this,
+                                                                            const std::shared_ptr<WriteCallback> &stream,
+                                                                            v_buff_size count, v_buff_size readOffset) {
   class WriteDataCoroutine : public oatpp::async::Coroutine<WriteDataCoroutine> {
```

Comment:
Looks like `WriteDataCoroutine` can be rewritten to something like this:

```cpp
  class WriteDataCoroutine : public oatpp::async::Coroutine<WriteDataCoroutine> {
   private:
    std::shared_ptr<BufferOutputStream> m_this;
    std::shared_ptr<oatpp::data::stream::WriteCallback> m_stream;
    data::buffer::InlineWriteData m_inlineData;
    v_buff_size m_count;
    v_buff_size m_readOffset;
   public:

    WriteDataCoroutine(const std::shared_ptr<BufferOutputStream>& _this,
                       const std::shared_ptr<oatpp::data::stream::WriteCallback>& stream,
                       v_buff_size count, v_buff_size readOffset)
        : m_this(_this)
        , m_stream(stream)
        , m_count(count)
        , m_readOffset(readOffset)
    {
      m_inlineData.set(m_this->m_data + m_readOffset, std::min(m_this->m_position - m_readOffset, m_count));
    }

    Action act() override {
      return m_stream->writeExactSizeDataAsyncInline(m_inlineData, finish());
    }

  };
```

There is no need in additional checks, plus inline data initialization can be moved to constructor.

Also, if `if (m_readOffset > m_this->m_position) ` then it's better to not start the coroutine:

```cpp
if(readOffset < m_position) {
  return WriteDataCoroutine::start(_this, stream, count, readOffset);
}
return nullptr; // do not start the coroutine
```

Reply:
Yes, now that I have some experience with `oatpp::async` I would have written it differently in the first place.